src/Mutator/OperationCategory.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * This code is licensed under the BSD 3-Clause License.
+ *
+ * Copyright (c) 2017, Maks Rafalko
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Mutator;
+
+final class OperationCategory
```

Comment:
Is it a replacement or an addition for `ProfileList`? What is the difference?

It's not immediately clear why do we have 2 categorization now.

Reply:
Let's take the example of the two following mutators:

```
$a + $b -> $b - $a
$a - $b -> $a + $b 
```

You can say the following for both:

- they are an orthogonal replacement
- they are mutators operating on an arithmetic operation
- they are dual

`ProfileList` is a different categorization which is orthogonal to `OperationCategory`: it helps to indicate on what kind of elements the mutator operates on, but there is no overlap whatsoever between `OperationCategory` and `ProfileList`.

Another aspect is that `OperationCategory` is purely internal whereas `ProfileList` helps to expose profiles which can be consumed by the user.

In other words: it's a different kind of categorization which is internal and serves more for double checking the mutators (e.g. if you see a semantic addition you know it's wrong) and providing basic remedies (for most of them you can infer it from its type)

Reply:
now it's more clear, thanks

* Are these "OperationCategory" and "remedies" some kind of well known definitions? If so, probably we can leave a couple of links so anyone can dig into the terminology
* If they are our definitions - does it makes sense to add everything you described to our docs or ADR?

Why is it needed: because in a week I will struggle to identify what `OperationCategory::*` really means why is it different / orthogonal with the `ProfileList` ;) WDYT?

Reply:
>Are these "OperationCategory" and "remedies" some kind of well known definitions?

Not at the moment. I'm keeping track of those terms on a separate file to lay down a nomenclature that I wish to submit on https://github.com/theofidry/awesome-mutation-testing (instead of like initially thought just infection) which would be peered review by the other mutation testing library authors.

The only well-agreed on terms here for now are those categories `semantic reduction`, `orthogonal replacement`, `neutral` and `semantic addition`. 

I'm not too satisfied with `OperationCategory`, could also be `MutatorCategory` instead... And `remedies` is one term that popped up as an idea of "fixing a mutant". So those two terms can be changed to something more fitting if you have any suggestion.

>Why is it needed: because in a week I (end even worse - someone who didn't see your explanation) will struggle to identify what OperationCategory::* really means why is it different / orthogonal with the ProfileList ;) WDYT?

I'm assuming you are talking of the category alone here as I hope the PR description highlights better the goal of `Definition`. IMO the category is important for internal quality and documentation in a number of ways:

- the category forces you, when you add or a review an undocumented mutator, to put some thoughts into what that mutator is doing. That process is necessary and make easier to spot smelly mutators (e.g. the ones which are a semantic addition or even the neutral ones which in our case should not be necessary yet)
- infer properties from the mutators for future improvements. 
  - For orthogonal replacement ones in peculiar where you can have the concept of `duality` and have auto-review checks to make sure when you declare a dual, that the other dual is declared
   - You can suggested "default" remedies based on the category, cf. the explanation in each category which I hope gives some hints on that

Last but not least, another goal is to improve the mutators taxonomy overall to make it as complete as possible in order to be able to better communicate and compare our mutators with other tools or review their efficiency.

Reply:
Renamed `OperationCategory` to `MutatorCategory` as it makes more sense for now I think. I hope we can improve the naming later on as we develop the taxonomy

Reply:
> I'm keeping track of those terms on a separate file to lay down a nomenclature that I wish to submit on https://github.com/theofidry/awesome-mutation-testing

IMO we need to refer to this file, otherwise nobody will understand what does it all mean

Reply:
I'll do so as soon as this file is public! I'm not publishing it yet because for now it's a bunch of notes and terms...